File: samples/video_encode/video_encode.py
import os
import sys
from threading import Thread
import vaststreamx as vsx

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from common.vencoder import Vencoder
import argparse


def read_frames(input_file, width, height):
    frames = []
    frames_num = 0
    with open(input_file, "rb") as file:
        while True:
            data = file.read((int)(width * height * 3 / 2))
            if len(data) == 0:
                break
            frames.append(data[:])
            frames_num += 1
    return frames, frames_num


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--codec_type",
        default="H264",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--width",
        default=0,
        type=int,
        help="frame width",
    )
    parser.add_argument(
        "--height",
        default=0,
        type=int,
        help="frame height",
    )
    parser.add_argument(
        "--frame_rate",
        default=30,
        type=int,
        help="frame rate",
    )
    parser.add_argument(
        "--input_file",
        default="",
        help="video file",
    )
    parser.add_argument(
        "--output_file",
        default="",
        help="output file",
    )

    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    codec_type = vsx.CodecType.CODEC_TYPE_H264
    if args.codec_type == "H264" or args.codec_type == "h264":
        codec_type = vsx.CodecType.CODEC_TYPE_H264
    elif args.codec_type == "H265" or args.codec_type == "h265":
        codec_type = vsx.CodecType.CODEC_TYPE_H265
    elif args.codec_type == "AV1" or args.codec_type == "av1":
        codec_type = vsx.CodecType.CODEC_TYPE_AV1
    else:
        print(f"undefined codec_type:{args.codec_type}")
        exit(-1)

    input_file = args.input_file
    output_file = args.output_file

    frames, frames_num = read_frames(input_file, args.width, args.height)

    vencoder = Vencoder(
        codec_type,
        args.device_id,
        frames,
        frames_num,
        args.width,
        args.height,
        vsx.ImageFormat.YUV_NV12,
        args.frame_rate,
    )

    def get_frame_thread():
        file = open(output_file, "wb")
        index = 0
        while True:
            try:
                video_data = vencoder.get_result()
                index += 1
                file.write(video_data)
            except ValueError as e:
                break

        file.close()

    recv_thread = Thread(target=get_frame_thread)
    recv_thread.start()

    while True:
        media_data = vencoder.get_test_data(False)
        if media_data is None:
            vencoder.stop()
            break
        else:
            vencoder.process(media_data)

    recv_thread.join()

File: samples/video_encode/video_encode.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <sys/stat.h>

#include <fstream>

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "common/vencoder.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("codec_type", '\0', "codec type", false, "H264");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<uint32_t>("width", '\0', "width", false, 0);
  args.add<uint32_t>("height", '\0', "height", false, 0);
  args.add<uint32_t>("frame_rate", '\0', "frame rate", false, 30);
  args.add<std::string>("input_file", '\0', "input file", false, "");
  args.add<std::string>("output_file", '\0', "output file", false, "");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  auto device_id = args.get<uint32_t>("device_id");
  auto width = args.get<uint32_t>("width");
  auto height = args.get<uint32_t>("height");
  auto frame_rate = args.get<uint32_t>("frame_rate");
  auto codec_type = args.get<std::string>("codec_type");

  auto input_file = args.get<std::string>("input_file");
  auto output_file = args.get<std::string>("output_file");

  CHECK(vsx::SetDevice(device_id) == 0)
      << "Failed to set device id: " << device_id;

  vsx::CodecType type;
  if (codec_type == "H264" || codec_type == "h264") {
    type = vsx::CODEC_TYPE_H264;
  } else if (codec_type == "H265" || codec_type == "h265") {
    type = vsx::CODEC_TYPE_HEVC;
  } else if (codec_type == "AV1" || codec_type == "av1") {
    type = vsx::CODEC_TYPE_AV1;
  } else {
    std::cerr << "undefined codec_type:" << codec_type << std::endl;
    return -1;
  }
  uint32_t frame_rate_numerator = frame_rate;
  uint32_t frame_rate_denominator = 1;
  vsx::Vencoder vencoder(type, device_id, input_file, width, height,
                         vsx::ImageFormat::YUV_NV12, frame_rate_numerator,
                         frame_rate_denominator);

  std::thread get_frame_thread([&]() {
    std::shared_ptr<vsx::DataManager> data;
    std::ofstream ofs(output_file, std::ios::binary);
    while (vencoder.GetResult(data)) {
      ofs.write(reinterpret_cast<const char*>(data->GetDataPtr()),
                data->GetDataSize());
    }
    ofs.close();
  });

  do {
    auto image = vencoder.GetTestData(false);
    if (image.Format() == vsx::ImageFormat::YUV_NV12 && image.GetDataPtr() &&
        image.GetDataBytes()) {
      vencoder.Process(image, false);
    } else {
      vencoder.Process(image, true);
      break;
    }
  } while (1);
  get_frame_thread.join();
  std::cout << "Write encode data to: " << output_file << std::endl;

  return 0;
}